src/Sia.Gateway/Filters/EventFilters.cs

```diff
@@ -20,27 +24,32 @@ public class EventFilters: Filters<Event>
         public const string KeyValueComparison = "\"{0}\":\"{1}\"";
         public const string KeyComparison = "\"{0}\":";
 
-        public override IQueryable<Event> Filter(IQueryable<Event> source)
+        public bool IsMatchFor(Data.Incidents.Models.Event toCompare)
         {
-            var working = source;
-
-            if (IncidentId.HasValue) working = working.Where(ev => ev.IncidentId == IncidentId);
-            if (EventTypes != null && EventTypes.Length > 0) working = working.Where(ev => EventTypes.Contains(ev.EventTypeId));
-            if (Occurred.HasValue) working = working.Where(ev => ev.Occurred == Occurred);
-            if (EventFired.HasValue) working = working.Where(ev => ev.EventFired == EventFired);
+            if (IncidentId.HasValue && toCompare.IncidentId != IncidentId.Value) return false;
+            if (EventTypes != null && EventTypes.Length > 0 && !EventTypes.Contains(toCompare.EventTypeId)) return false;
+            if (Occurred.HasValue && toCompare.Occurred != Occurred.Value) return false;
+            if (EventFired.HasValue && toCompare.EventFired != EventFired.Value) return false;
 
             if (!String.IsNullOrEmpty(DataKey))
             {
-                var workingCompare = String.IsNullOrEmpty(DataValue)
-                    ? String.Format(KeyComparison, DataKey)
-                    : String.Format(KeyValueComparison, new string[] { DataKey, DataValue });
-                working = working.Where(obj => obj.Data.Contains(workingCompare));
+                if (String.IsNullOrEmpty(DataValue))
```

Comment:
API consumers have no control over the format of the JSON string we store in the database; they post an object and Newtonsoft JSONConvert serializes it to a string, which should result in consistent formatting for these columns in the database. We may want to validate our serializer settings just to be sure.

We are unable to do deserialization in these comparison methods because they're executed within the SQL server engine rather than on the Gateway instance itself, and Entity Framework's Linq provider does not have a way to convert the JSONConvert functions to (I assume) IL that can be executed within the SQL Server engine. This is the same reason we can't use the null-coalesce operator in these functions. We would need to return a (potentially huge) superset of the data to the Gateway and do the filtering in memory if we wanted to test against the deserialized version.